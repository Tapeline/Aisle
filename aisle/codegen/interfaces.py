from abc import ABC, abstractmethod
from typing import Callable, Mapping

from aisle.analyser.entities.project import Project


type FileGenerator = Callable[[], str]


class AbstractProjectGenerator(ABC):
    """ABC for project code generators."""

    file_extension: str = ""

    @abstractmethod
    def __init__(self, project: Project) -> None:
        """Create project generator."""

    @property
    @abstractmethod
    def file_generators(self) -> Mapping[str, FileGenerator]:
        """Get generators"""
