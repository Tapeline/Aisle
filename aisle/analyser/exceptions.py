from aisle.parser.nodes.base import Node


class VisitMethodNotFoundError(AttributeError):
    def __init__(self, class_name: str, obj):  # pragma: no cover
        super().__init__(obj=obj, name=f"Method visit_{class_name}")
        self.class_name = class_name


class AnalyserException(Exception):
    default_message: str = "Invalid node"

    def __init__(
            self,
            node: Node,
            message: str | None = None,
            **params,
    ):
        self.node = node
        self.message = message or self.default_message
        for param_k, param_v in params.items():
            setattr(self, param_k, param_v)

    def formatted_message(self, source_code: str):
        lines = source_code.splitlines()
        formatted = self.message.format(node=self.node)
        return (
            f"Exception while analysing node {self.node}\n"
            f"At line: {self.node.line}\n"
            f"{self.node.line} |  {lines[self.node.line - 1]}\n"
            f"{formatted}"
        )


class NoProjectDefinedException(AnalyserException):
    default_message: str = (
        "Found {node.__class__.__name__} node, but no project was defined. "
        "Define a project first!"
    )


class DuplicateProjectDefinitionException(AnalyserException):
    default_message: str = (
        "Found project definition, but project was already defined."
    )


class UnmatchedProjectAndScopeNameException(AnalyserException):
    default_message: str = (
        "Found scope with '{scope_name}', "
        "but project is called '{project_name}'"
    )


class UnmatchedScopeAndEntityTypeException(AnalyserException):
    default_message: str = (
        "In {scope_type} scope, there was found "
        "an entity of {entity_type} type"
    )
