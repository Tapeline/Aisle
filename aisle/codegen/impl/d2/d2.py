import textwrap
from collections.abc import Mapping
from typing import assert_never

from aisle.analyser.entities.containers import ServiceEntity
from aisle.analyser.entities.context import ActorEntity, SystemEntity
from aisle.analyser.entities.deployment import DeploymentEntity
from aisle.analyser.entities.links import Link
from aisle.analyser.entities.project import Project
from aisle.codegen.impl.d2.text_utils import safe_name
from aisle.codegen.interfaces import AbstractProjectGenerator, FileGenerator
from aisle.parser.nodes.links import LinkType


class _D2SafeNameStorage:
    """Automatically assigns safe names according to d2."""

    def __init__(self) -> None:
        """Create storage."""
        self._names: dict[str, str] = {}

    def __getitem__(self, unsafe_name: str) -> str:
        """Get safe name or generate if not present."""
        if unsafe_name not in self._names:
            self._names[unsafe_name] = safe_name(unsafe_name)
        return self._names[unsafe_name]

    @property
    def names(self) -> dict[str, str]:
        return self._names


class _CodeGenerator:
    def __init__(self, project: Project) -> None:
        """Create generator."""
        self.project = project
        self._safe_names = _D2SafeNameStorage()

    def get_main_map(self) -> str:
        """Generate main view."""
        strings: list[str] = []
        strings.extend(
            self._gen_actor(actor)
            for actor in self.project.get_actors()
        )
        strings.extend(
            self._gen_system(system)
            for system in self.project.get_systems()
        )
        strings.append(self._gen_links())
        return "\n\n".join(strings)

    def get_deployment_map(self) -> str:
        return "\n\n".join(self._gen_deployments())

    def _gen_system(self, system: SystemEntity) -> str:
        sys_name = self._safe_names[system.name]
        sys_desc = textwrap.indent(system.description, "  ")
        short_desc = textwrap.indent(_gen_short_desc(system), "  ")
        system_code = (
            f"{sys_name}: |md\n"
            f"  ## {system.name}\n"
            f"{short_desc}\n\n"
            f"{sys_desc}\n"
            "| {\n"
            "  shape: rectangle\n"
            "  label.near: bottom-left\n"
            "}\n"
        )
        services = [
            self._gen_service(system, service)
            for service in self.project.get_services_of_system(system.name)
        ]
        return "\n\n".join([system_code, *services])

    def _gen_service(
            self, system: SystemEntity, service: ServiceEntity
    ) -> str:
        svc_name = self._safe_names[service.name]
        sys_name = self._safe_names[system.name]
        svc_desc = textwrap.indent(service.description, "  ")
        short_desc = textwrap.indent(_gen_short_desc(service), "  ")
        svc_shape = _get_service_shape(service)
        return (
            f"{sys_name}.{svc_name}: |md\n"
            f"  ## {service.name}\n"
            f"{short_desc}\n\n"
            f"{svc_desc}\n"
            "| {\n"
            f"  shape: {svc_shape}\n"
            "}\n"
        )

    def _gen_actor(self, actor: ActorEntity) -> str:
        actor_name = self._safe_names[actor.name]
        actor_desc = textwrap.indent(actor.description, "  ")
        short_desc = textwrap.indent(_gen_short_desc(actor), "  ")
        return (
            f"{actor_name}: |md\n"
            f"  ## {actor.name}\n"
            f"{short_desc}\n\n"
            f"{actor_desc}\n"
            "| {\n"
            "  shape: c4-person\n"
            "}\n"
        )

    def _get_d2_names_map(self) -> Mapping[str, str]:
        names = self._safe_names.names.copy()
        for system in self.project.get_systems():
            for service in self.project.get_services_of_system(system.name):
                names[service.name] = (
                    f"{names[system.name]}.{self._safe_names[service.name]}"
                )
        return names

    def _gen_links(self) -> str:
        links: list[str] = []
        names = self._get_d2_names_map()
        for actor in self.project.get_actors():
            links.extend(
                _gen_link(
                    names[actor.name], names[link.to], link
                )
                for link in actor.links
            )
        for system in self.project.get_systems():
            links.extend(
                _gen_link(
                    names[system.name], names[link.to], link
                )
                for link in system.links
            )
            for service in self.project.get_services_of_system(system.name):
                links.extend(
                    _gen_link(
                        names[service.name], names[link.to], link
                    )
                    for link in service.links
                )
        return "\n".join(links)

    def _gen_deployments(self) -> list[str]:
        return [
            "\n\n".join(
                _gen_deployment_rec(self._safe_names, [], deployment)
            )
            for deployment in self.project.get_deployments()
        ]


def _gen_deployment_rec(  # noqa: WPS210
        safe_names: _D2SafeNameStorage,
        prefix: list[str],
        deployment: DeploymentEntity
) -> list[str]:
    dep_name = safe_names[deployment.name]
    dep_desc = textwrap.indent(deployment.description, "  ")
    short_desc = textwrap.indent(_gen_short_desc(deployment), "  ")
    qualname = ".".join([*prefix, dep_name])
    dep_code = (
        f"{qualname}: |md\n"
        f"  ## {deployment.name}\n"
        f"{short_desc}\n\n"
        f"{dep_desc}\n"
        "| {\n"
        "  shape: rectangle\n"
        "  label.near: bottom-left\n"
        "}\n"
    )
    deployments = [dep_code]
    deployments.extend(
        (
            f"{qualname}.{safe_names[service.service_name]}: |md\n"
            f"  ### {service.service_name}\n"
            f"  deploy as {service.deploy_as}\n"
            "| {\n"
            "  shape: rectangle\n"
            "}\n"
        )
        for service in deployment.deploys
    )
    for sub_dep in deployment.inner_entities:
        deployments.extend(
            _gen_deployment_rec(safe_names, [*prefix, dep_name], sub_dep)
        )
    return deployments


def _gen_short_desc(
        entity: SystemEntity | ServiceEntity | ActorEntity | DeploymentEntity
) -> str:
    if isinstance(entity, SystemEntity):
        is_ext = " (external)" if entity.is_external else ""
        return f"[Software System{is_ext}]"
    if isinstance(entity, ServiceEntity):
        is_ext = " (external)" if entity.is_external else ""
        return f"[Container{is_ext}: {entity.tech}]"
    if isinstance(entity, ActorEntity):
        return "[Actor]"
    if isinstance(entity, DeploymentEntity):
        return "[Deployment]"
    assert_never(entity)


def _get_service_shape(service: ServiceEntity) -> str:
    if "db" in service.tags or "database" in service.tags:
        return "cylinder"
    return "rectangle"


def _gen_link(from_name: str, to_name: str, link: Link) -> str:
    link_t = {
        LinkType.OUTGOING: "->",
        LinkType.INCOMING: "<-",
        LinkType.BIDIRECTIONAL: "<->",
        LinkType.NON_DIRECTED: "<->",
    }[link.type]
    link_desc = []
    if link.over:
        link_desc.append(f"[{link.over}]")
    if link.description:
        link_desc.append(repr(link.description).strip("'"))
    link_desc_s = " ".join(link_desc)
    return f'{from_name} {link_t} {to_name}: "{link_desc_s}"'


class D2ProjectGenerator(AbstractProjectGenerator):
    """D2lang code generator."""

    file_extension = "d2"

    def __init__(self, project: Project) -> None:
        """Create generator."""
        self.project = project
        self._cg = _CodeGenerator(project)

    @property
    def file_generators(self) -> Mapping[str, FileGenerator]:
        """Get generators."""
        return {
            "main": self.generate_main,
            "deployments": self.generate_deployments
        }

    def generate_main(self) -> str:
        """Generate code."""
        return self._cg.get_main_map()

    def generate_deployments(self) -> str:
        """Generate code for deployments."""
        return self._cg.get_deployment_map()
