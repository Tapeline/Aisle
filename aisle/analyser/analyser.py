from collections.abc import Sequence
from dataclasses import dataclass

from aisle.analyser.entities.containers import ServiceEntity
from aisle.analyser.entities.context import ActorEntity, SystemEntity
from aisle.analyser.entities.deployment import DeploymentEntity, ServiceDeployment
from aisle.analyser.entities.links import Link
from aisle.analyser.entities.project import Project
from aisle.analyser.entities.styling import LegendStyling, StylingAttributes
from aisle.analyser.exceptions import *
from aisle.parser.nodes.attribute import AttrNode
from aisle.parser.nodes.base import Node
from aisle.parser.nodes.entity import ProjectDefNode, EntityNode, EntityType, DeployNode
from aisle.parser.nodes.legend import LegendDeclarationNode
from aisle.parser.nodes.links import LinkCollectionNode
from aisle.parser.nodes.scope import ScopeNode, ScopeType
from aisle.parser.nodes.text import TextNode


class AnalyserBase:
    def _visit(self, node: Node):
        class_name = node.__class__.__name__
        self._pre_visit(node)
        try:
            getattr(self, f"_visit_{class_name}")(node)
        except AttributeError as error:
            raise VisitMethodNotFoundError(class_name, self) from error

    def _pre_visit(self, node: Node):
        ...


@dataclass
class Scope:
    type: ScopeType
    name: str


class Analyser(AnalyserBase):
    def __init__(self, nodes: Sequence[Node]):
        self.project: Project | None = None
        self.scope: Scope | None = None
        self.nodes = nodes

    def analyse(self) -> Project:
        for node in self.nodes:
            self._visit(node)
        if self.project is None:
            raise NoProjectDefinedException(Node(0), message="No project found")
        return self.project

    def _pre_visit(self, node: Node):
        if self.project is None and not isinstance(node, ProjectDefNode):
            raise NoProjectDefinedException(node)

    def _visit_ProjectDefNode(self, node: ProjectDefNode):
        if self.project is not None:
            raise DuplicateProjectDefinitionException(node)
        description = "\n".join(node.description)
        self.project = Project(
            name=node.name,
            description=description,
            namespace={},
            styling=[],
            comments=[]
        )

    def _visit_ScopeNode(self, node: ScopeNode):
        if self.project.name != node.scope_name:
            raise UnmatchedProjectAndScopeNameException(
                node,
                scope_name=node.scope_name,
                project_name=self.project.name
            )
        self.scope = Scope(node.scope_type, node.scope_name)

    def _ensure_entity_applicable(self, entity: EntityNode):
        if self.scope is None:
            raise UnmatchedScopeAndEntityTypeException(
                entity,
                scope_type=None,
                entity_type=entity.type.value
            )
        s_type = self.scope.type
        e_type = entity.type
        if not (
            s_type == ScopeType.CONTEXT and e_type == EntityType.ACTOR
            or s_type == ScopeType.CONTEXT and e_type == EntityType.SYSTEM
            or s_type == ScopeType.CONTAINERS and e_type == EntityType.SERVICE
            or s_type == ScopeType.DEPLOYMENT and e_type == EntityType.DEPLOYMENT
        ):
            raise UnmatchedScopeAndEntityTypeException(
                entity,
                scope_type=s_type.value,
                entity_type=e_type.value
            )

    def _visit_EntityNode(self, node: EntityNode):
        self._ensure_entity_applicable(node)
        match node.type:
            case EntityType.ACTOR:
                self._create_actor(node)
            case EntityType.SYSTEM:
                self._create_system(node)
            case EntityType.SERVICE:
                self._create_service(node)
            case EntityType.DEPLOYMENT:
                self._create_deployment(node)

    def _evaluate_entity_body(
            self, body: Sequence[Node]
    ) -> tuple[str, list[Link], dict[str, str]]:
        description = "\n".join(
            node.text for node in body if isinstance(node, TextNode)
        )
        attrs = {
            node.name: node.value
            for node in body if isinstance(node, AttrNode)
        }
        links = []
        for node in body:
            if isinstance(node, LinkCollectionNode):
                links.extend(node.links)
        links = [
            Link(
                type=link.type,
                to=link.to,
                over=link.over,
                description="\n".join(link.description)
            )
            for link in links
        ]
        return description, links, attrs

    def _create_actor(self, node: EntityNode):
        description, links, _ = self._evaluate_entity_body(node.body)
        entity = ActorEntity(
            name=node.name,
            description=description,
            links=links,
            tags=list(node.tags)
        )
        self.project.namespace[entity.name] = entity

    def _create_system(self, node: EntityNode):
        description, links, _ = self._evaluate_entity_body(node.body)
        entity = SystemEntity(
            name=node.name,
            description=description,
            links=links,
            tags=list(node.tags)
        )
        self.project.namespace[entity.name] = entity

    def _create_service(self, node: EntityNode):
        description, links, attrs = self._evaluate_entity_body(node.body)
        entity = ServiceEntity(
            name=node.name,
            description=description,
            links=links,
            tech=attrs.get("tech"),
            system=attrs.get("system"),
            tags=list(node.tags)
        )
        self.project.namespace[entity.name] = entity

    def _create_deployment(self, node: EntityNode):
        deploys = [
            ServiceDeployment(deploy.target, deploy.deploy_as)
            for deploy in node.body if isinstance(deploy, DeployNode)
        ]
        description, _, _ = self._evaluate_entity_body(node.body)
        entity = DeploymentEntity(
            name=node.name,
            description=description,
            deploys=deploys,
            tags=list(node.tags),
        )
        self.project.namespace[entity.name] = entity

    def _visit_TextNode(self, node: TextNode):
        self.project.comments.append(node.text)

    def _visit_LegendDeclarationNode(self, node: LegendDeclarationNode):
        description, _, attrs = self._evaluate_entity_body(node.body)
        style = LegendStyling(
            selector=node.selector,
            description=description,
            attrs=StylingAttributes(**attrs)
        )
        self.project.styling.append(style)
