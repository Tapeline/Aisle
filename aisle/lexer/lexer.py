"""Contains lexer implementation."""

from collections.abc import Sequence

from aisle.lexer.exceptions import (
    UnexpectedCharacterException,
    StringNotClosedException,
    IncompleteUnicodeQuadException,
)
from aisle.lexer.tokens import (
    Token,
    TokenType,
    RESERVED_STRINGS,
    KEYWORDS,
)


class Lexer:
    def __init__(
            self,
            source: str,
            tab_size: int = 4,
    ) -> None:
        self._src = _norm_newlines(source)
        self._lines = self._src.splitlines()
        self._i = 0
        self._start = 0
        self._line = 1
        self._line_start = 0
        self._tokens = []
        self._tab_size = tab_size

    @property
    def _at_end(self):
        """Check if reached end of source."""
        return self._i >= len(self._src)

    def _next(self) -> str | None:
        if self._i >= len(self._src):
            return None
        self._i += 1
        return self._src[self._i - 1]

    @property
    def _current(self) -> str | None:
        return None if self._at_end else self._src[self._i]

    def _add_token(
            self,
            token_type: TokenType,
            lexeme: str | None = None
    ) -> None:
        token = Token(
            token_type,
            lexeme or self._src[self._start:self._i],
            self._line
        )
        self._start = self._i
        self._tokens.append(token)

    def _match(self, pattern: str) -> str | None:
        if self._at_end:
            return None
        src = self._src[self._i:]
        if src.startswith(pattern):
            self._i += len(pattern)
            return pattern
        return None

    def _error(self, exc_type, *args, **kwargs) -> None:
        raise exc_type(
            *args,
            **kwargs,
            source=self._src,
            line=self._line
        )

    def scan(self) -> Sequence[Token]:
        while not self._at_end:
            self._scan_token()
        return self._tokens

    def _scan_token(self) -> None:
        c = self._next()
        match c:
            case "=":
                self._add_token(TokenType.ASSIGN)
            case ":":
                self._add_token(TokenType.COLON)
            case "(":
                self._add_token(TokenType.LPAR)
            case ")":
                self._add_token(TokenType.RPAR)
            case "[":
                self._add_token(TokenType.LBRACKET)
            case "]":
                self._add_token(TokenType.RBRACKET)
            case "-":
                if self._match("->"):
                    self._add_token(TokenType.ARROW_R)
                elif self._match("--"):
                    self._add_token(TokenType.ARROW_NO_DIR)
                else:
                    self._error(UnexpectedCharacterException, self._current)
            case "<":
                if self._match("->"):
                    self._add_token(TokenType.ARROW_BI_DIR)
                elif self._match("--"):
                    self._add_token(TokenType.ARROW_L)
                else:
                    self._error(UnexpectedCharacterException, self._current)
            case "\n":
                self._add_token(TokenType.NEWLINE)
                self._line += 1
                self._line_start = self._i
            case " ":
                if self._match(" " * (self._tab_size - 1)):
                    self._add_token(TokenType.INDENT)
                else:
                    self._start = self._i
            case "\t":
                self._add_token(TokenType.INDENT)
            case '"':
                self._scan_string()
            case _:
                self._i -= 1
                self._scan_text()

    def _scan_string(self) -> None:
        text = ""
        while not self._at_end and self._current != '"':
            if self._current == "\\":
                text += self._scan_escape_seq()
            else:
                text += self._current
                self._i += 1
        if not (not self._at_end and self._current == '"'):
            self._error(StringNotClosedException)
        self._i += 1
        self._add_token(TokenType.TEXT, text)

    def _scan_escape_seq(self) -> str:
        self._next()
        control = self._next()
        match control:
            case "r":
                return "\r"
            case "f":
                return "\f"
            case "n":
                return "\n"
            case "t":
                return "\t"
            case '"':
                return '"'
            case "\\":
                return "\\"
            case "u":
                code = int(self._scan_unicode_quad(), 16)
                return chr(code)
        self._error(UnexpectedCharacterException, char=control)

    def _scan_unicode_quad(self) -> str:
        code = ""
        for _ in range(4):
            c = self._next()
            if c is None:
                self._error(IncompleteUnicodeQuadException)
            if c not in "0123456789abcdefABCDEF":
                self._error(IncompleteUnicodeQuadException)
            code += c
        return code

    def _scan_text(self) -> None:
        subindent_level = self._i - self._line_start
        text = ""
        while not self._at_end:
            if self._current == "\\":
                text += self._scan_escape_seq()
            elif self._current == "\n":
                self._line += 1
                if not self._next_line_on_same_level(subindent_level):
                    break
                text += self._current
                text += " "*subindent_level  # Retain all indentation
                self._i += subindent_level + 1
            else:
                text += self._current
                self._i += 1
            if _get_reserved_string_at_end(text):
                break
        reserved = _get_reserved_string_at_end(text)
        if reserved and len(reserved) != len(text):
            self._i -= len(reserved)
            text = text.removesuffix(reserved)

        # Return all extra indentation and newline symbols back
        stripped_text = text.strip()
        suffix_start = text.index(stripped_text) + len(stripped_text)
        suffix_length = len(text) - suffix_start
        self._i -= suffix_length
        # Update line count accordingly
        self._line -= text[suffix_start:].count("\n")
        # Remove unnecessary indentation
        text = _remove_indentation(text, subindent_level)

        if text in KEYWORDS:
            self._add_token(TokenType.KEYWORD, text.strip())
        else:
            self._add_token(TokenType.TEXT, text.strip())

    def _next_line_on_same_level(self, subindent_level: int) -> bool:
        return self._lines[self._line - 1].startswith(" " * subindent_level)


def _get_reserved_string_at_end(text: str) -> str | None:
    if text in RESERVED_STRINGS:
        return text
    for string in RESERVED_STRINGS:
        if _should_be_split(text, string):
            return string
    return None


def _should_be_split(string: str, reserved: str) -> bool:
    if not string.endswith(reserved):
        return False
    trunk = string.removesuffix(reserved)
    if not reserved.isalnum():
        return True
    return trunk[-1] in "\n\r\t "


def _norm_newlines(text: str) -> str:
    """Replace all CRLF with LF."""
    return text.replace("\r\n", "\n").replace("\r", "\n")


def _remove_indentation(text: str, amount: int) -> str:
    lines = text.splitlines()
    prefix = " "*amount
    for i, line in enumerate(lines):
        lines[i] = line.removeprefix(prefix)
    return "\n".join(lines)

"""
tech = fdffd

"""