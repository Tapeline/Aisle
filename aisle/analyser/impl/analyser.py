from collections.abc import Sequence
from dataclasses import dataclass
from types import MappingProxyType

from aisle.analyser.entities.containers import ServiceEntity
from aisle.analyser.entities.context import ActorEntity, SystemEntity
from aisle.analyser.entities.deployment import (
    DeploymentEntity,
    ServiceDeployment,
)
from aisle.analyser.entities.links import Link
from aisle.analyser.entities.project import Namespace, Project
from aisle.analyser.entities.styling import LegendStyling, StylingAttributes
from aisle.analyser.exceptions import (
    DuplicateProjectDefinitionError,
    NoProjectDefinedError,
    UnmatchedProjectAndScopeNameError,
    UnmatchedScopeAndEntityTypeError,
    VisitMethodNotFoundError,
)
from aisle.analyser.interfaces import AbstractAnalyser
from aisle.parser.nodes.attribute import AttrNode
from aisle.parser.nodes.base import Node
from aisle.parser.nodes.entity import (
    DeployNode,
    EntityNode,
    EntityType,
    ProjectDefNode,
)
from aisle.parser.nodes.legend import LegendDeclarationNode
from aisle.parser.nodes.links import LinkCollectionNode, LinkNode
from aisle.parser.nodes.scope import ScopeNode, ScopeType
from aisle.parser.nodes.text import TextNode


class _AnalyserBase:
    def _visit(self, node: Node) -> None:
        class_name = node.__class__.__name__
        self._pre_visit(node)
        try:
            getattr(self, f"_visit_{class_name}")(node)
        except AttributeError as error:  # pragma: no cover
            raise VisitMethodNotFoundError(class_name, self) from error

    def _pre_visit(self, node: Node) -> None:
        ...  # pragma: no cover


@dataclass
class Scope:
    """Analyser scope."""

    type: ScopeType
    name: str


_ALLOWED_ENTITY_TYPES = MappingProxyType(
    {
        ScopeType.CONTEXT: {EntityType.ACTOR, EntityType.SYSTEM},
        ScopeType.CONTAINERS: {EntityType.SERVICE},
        ScopeType.DEPLOYMENT: {EntityType.DEPLOYMENT},
    }
)


class Analyser(AbstractAnalyser, _AnalyserBase):
    """Default analyser impl."""

    def __init__(self, nodes: Sequence[Node]):
        """Create analyser."""
        self.project: Project | None = None
        self.scope: Scope | None = None
        self.nodes = nodes

    def analyse(self) -> Project:
        """Analyse AST and create project."""
        for node in self.nodes:
            self._visit(node)
        if self.project is None:
            raise NoProjectDefinedError(
                Node(0),
                message="No project found"
            )
        return self.project

    def _pre_visit(self, node: Node) -> None:
        if self.project is None and not isinstance(node, ProjectDefNode):
            raise NoProjectDefinedError(node)

    def _visit_ProjectDefNode(self, node: ProjectDefNode) -> None:
        if self.project is not None:
            raise DuplicateProjectDefinitionError(node)
        description = "\n".join(node.description)
        self.project = Project(
            name=node.name,
            description=description,
            namespace=Namespace(),
            styling=[],
            comments=[]
        )

    def _visit_ScopeNode(self, node: ScopeNode) -> None:
        if self.project.name != node.scope_name:
            raise UnmatchedProjectAndScopeNameError(
                node,
                scope_name=node.scope_name,
                project_name=self.project.name
            )
        self.scope = Scope(node.scope_type, node.scope_name)

    def _ensure_entity_applicable(  # pragma: no cover
            self,
            entity: EntityNode
    ) -> None:
        if self.scope is None:
            raise UnmatchedScopeAndEntityTypeError(
                entity,
                scope_type=None,
                entity_type=entity.type.value
            )
        if entity.type not in _ALLOWED_ENTITY_TYPES[self.scope.type]:
            raise UnmatchedScopeAndEntityTypeError(
                entity,
                scope_type=self.scope.type.value,
                entity_type=entity.type.value
            )

    def _visit_EntityNode(self, node: EntityNode) -> None:
        self._ensure_entity_applicable(node)
        match node.type:
            case EntityType.ACTOR:
                self._create_actor(node)
            case EntityType.SYSTEM:
                self._create_system(node)
            case EntityType.SERVICE:
                self._create_service(node)
            case EntityType.DEPLOYMENT:
                self._create_deployment(node)

    def _evaluate_entity_body(
            self, body: Sequence[Node]
    ) -> tuple[str, list[Link], dict[str, str]]:
        description = "\n".join(
            node.text for node in body if isinstance(node, TextNode)
        )
        attrs = {
            node.name: node.value
            for node in body if isinstance(node, AttrNode)
        }
        collected_links: list[LinkNode] = []
        for node in body:
            if isinstance(node, LinkCollectionNode):
                collected_links.extend(node.links)
        links = [
            Link(
                type=link.type,
                to=link.to,
                over=link.over,
                description="\n".join(link.description)
            )
            for link in collected_links
        ]
        return description, links, attrs

    def _evaluate_deployment_body(
            self,
            body: Sequence[Node]
    ) -> tuple[str, list[ServiceDeployment], list[DeploymentEntity]]:
        description = "\n".join(
            node.text for node in body if isinstance(node, TextNode)
        )
        svc_deployments = [
            ServiceDeployment(node.target, node.deploy_as)
            for node in body if isinstance(node, DeployNode)
        ]
        inner_deployments = []
        for node in body:
            if not (
                    isinstance(node, EntityNode)
                    and node.type == EntityType.DEPLOYMENT
            ):
                continue
            i_desc, i_svc, i_inner = self._evaluate_deployment_body(node.body)
            inner_deployments.append(
                DeploymentEntity(
                    name=node.name,
                    description=i_desc,
                    tags=list(node.tags),
                    is_external=node.is_external,
                    deploys=i_svc,
                    inner_entities=i_inner
                )
            )
        return description, svc_deployments, inner_deployments

    def _create_actor(self, node: EntityNode) -> None:
        description, links, _ = self._evaluate_entity_body(node.body)
        entity = ActorEntity(
            name=node.name,
            description=description,
            links=links,
            tags=list(node.tags)
        )
        self.project.namespace[entity.name] = entity

    def _create_system(self, node: EntityNode) -> None:
        description, links, _ = self._evaluate_entity_body(node.body)
        entity = SystemEntity(
            name=node.name,
            description=description,
            links=links,
            tags=list(node.tags),
            is_external=node.is_external
        )
        self.project.namespace[entity.name] = entity

    def _create_service(self, node: EntityNode) -> None:
        description, links, attrs = self._evaluate_entity_body(node.body)
        entity = ServiceEntity(
            name=node.name,
            description=description,
            links=links,
            tech=attrs.get("tech"),
            system=attrs.get("system"),
            tags=list(node.tags),
            is_external=node.is_external
        )
        self.project.namespace[entity.name] = entity

    def _create_deployment(self, node: EntityNode) -> None:
        description, svc_deploys, inner_entities = (
            self._evaluate_deployment_body(node.body)
        )
        entity = DeploymentEntity(
            name=node.name,
            description=description,
            deploys=svc_deploys,
            tags=list(node.tags),
            is_external=node.is_external,
            inner_entities=inner_entities
        )
        self.project.namespace[entity.name] = entity

    def _visit_TextNode(self, node: TextNode) -> None:
        self.project.comments.append(node.text)

    def _visit_LegendDeclarationNode(
            self,
            node: LegendDeclarationNode
    ) -> None:
        description, _, attrs = self._evaluate_entity_body(node.body)
        style = LegendStyling(
            selector=node.selector,
            description=description,
            attrs=StylingAttributes(**attrs)
        )
        self.project.styling.append(style)
