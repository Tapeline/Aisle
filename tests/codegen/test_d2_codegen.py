from pathlib import Path

from aisle.analyser.impl.analyser import Analyser
from aisle.codegen.impl.d2.d2 import D2ProjectGenerator
from aisle.codegen.interfaces import AbstractProjectGenerator
from aisle.lexer.impl.lexer import Lexer
from aisle.parser.impl.parser import Parser


def _get_generator(path: str) -> AbstractProjectGenerator:
    src = Path(path).read_text()
    tokens = Lexer(src).scan()
    nodes = Parser(src, tokens).parse()
    project = Analyser(nodes).analyse()
    return D2ProjectGenerator(project)


def test_main_codegen(snapshot):
    """Test that context and container layers generate correctly."""
    generator = _get_generator("./tests/fixtures/test_1_a.aisle")
    assert generator.file_generators["main"]() == snapshot


def test_deployments_codegen(snapshot):
    """Test that deployment layer generates correctly."""
    generator = _get_generator("./tests/fixtures/test_1_a.aisle")
    assert generator.file_generators["deployments"]() == snapshot


def test_link_directions_codegen(snapshot):
    """Test that different link directions work correctly."""
    generator = _get_generator("./tests/fixtures/test_2.aisle")
    assert generator.file_generators["main"]() == snapshot


def test_name_clashes_codegen(snapshot):
    """Test that clashing names are resolved correctly."""
    generator = _get_generator("./tests/fixtures/test_3.aisle")
    assert generator.file_generators["main"]() == snapshot
