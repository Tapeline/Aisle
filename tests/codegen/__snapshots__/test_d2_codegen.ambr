# serializer version: 1
# name: test_deployments_codegen
  '''
  VPS Deployment: |md
    ## VPS Deployment
    [Deployment]
  
    Docker containers with Docker compose
  | {
    shape: rectangle
    label.near: bottom-left
  }
  
  
  VPS Deployment.Docker compose: |md
    ## Docker compose
    [Deployment]
  
  
  | {
    shape: rectangle
    label.near: bottom-left
  }
  
  
  VPS Deployment.Docker compose.Frontend Service: |md
    ### Frontend Service
    deploy as Docker container
  | {
    shape: rectangle
  }
  
  
  VPS Deployment.Docker compose.Backend Service: |md
    ### Backend Service
    deploy as Docker container
  | {
    shape: rectangle
  }
  
  
  VPS Deployment.Docker compose.DB Service: |md
    ### DB Service
    deploy as Docker container
  | {
    shape: rectangle
  }
  
  
  Provided Metrics: |md
    ## Provided Metrics
    [Deployment]
  
    Metrics Service is already provided
  | {
    shape: rectangle
    label.near: bottom-left
  }
  
  
  Provided Metrics.Metrics Service: |md
    ### Metrics Service
    deploy as Cloud provider
  | {
    shape: rectangle
  }
  
  '''
# ---
# name: test_link_directions_codegen
  '''
  User: |md
    ## User
    [Actor]
  
  
  | {
    shape: c4-person
  }
  
  
  System: |md
    ## System
    [Software System]
  
  
  | {
    shape: rectangle
    label.near: bottom-left
  }
  
  
  User -> System: ""
  User <- System: ""
  User <-> System: ""
  User <-> System: ""
  '''
# ---
# name: test_main_codegen
  '''
  User: |md
    ## User
    [Actor]
  
    Uses blogging app. Reads and creates posts
  | {
    shape: c4-person
  }
  
  
  Admin: |md
    ## Admin
    [Actor]
  
    Blog moderator
  | {
    shape: c4-person
  }
  
  
  Blogging System: |md
    ## Blogging System
    [Software System]
  
    Stores and manages blogs
  | {
    shape: rectangle
    label.near: bottom-left
  }
  
  
  Blogging System.Frontend Service: |md
    ## Frontend Service
    [Container: Vite+Vue]
  
    UI
  | {
    shape: rectangle
  }
  
  
  Blogging System.Backend Service: |md
    ## Backend Service
    [Container: Litestar, sqlalchemy]
  
    Provides API
    Maybe split into microservices later
  | {
    shape: rectangle
  }
  
  
  Blogging System.DB Service: |md
    ## DB Service
    [Container: Postgres]
  
  
  | {
    shape: cylinder
  }
  
  
  Metrics System: |md
    ## Metrics System
    [Software System (external)]
  
    Collects site metrics
    We will use Yandex Metrics
  | {
    shape: rectangle
    label.near: bottom-left
  }
  
  
  Metrics System.Metrics Service: |md
    ## Metrics Service
    [Container (external): None]
  
  
  | {
    shape: rectangle
  }
  
  
  User -> Blogging System: "Uses"
  Admin -> Blogging System: "Uses"
  Admin -> Metrics System: "Reviews post metrics"
  Blogging System -> Metrics System: ""
  Blogging System.Frontend Service -> Blogging System.Backend Service: "[HTTP] Send API requests"
  Blogging System.Frontend Service <- User: ""
  Blogging System.Frontend Service <- Admin: ""
  Blogging System.Backend Service -> Metrics System.Metrics Service: "[HTTP] Send metrics info on each request"
  Blogging System.Backend Service -> Blogging System.DB Service: "Stores info"
  '''
# ---
# name: test_name_clashes_codegen
  '''
  Plus_c7da77da: |md
    ## Plus+
    [Actor]
  
  
  | {
    shape: c4-person
  }
  
  
  Plus: |md
    ## Plus
    [Software System]
  
  
  | {
    shape: rectangle
    label.near: bottom-left
  }
  
  
  
  '''
# ---
